Kernel/FileSystem/VirtualFileSystem.cpp

```diff
@@ -163,41 +265,42 @@ ErrorOr<void> VirtualFileSystem::unmount(Custody& mountpoint_custody)
-                    m_file_systems_list.with([&](auto& list) {
-                        list.remove(*fs);
-                    });
-                    if (fs->is_file_backed()) {
-                        dbgln("VirtualFileSystem: Unmounting file backed file system {} for the last time...", fs->fsid());
-                        auto& file_backed_fs = static_cast<FileBackedFileSystem&>(*fs);
-                        m_file_backed_file_systems_list.with([&](auto& list) {
-                            list.remove(file_backed_fs);
+    return m_file_backed_file_systems_list.with_exclusive([&](auto& file_backed_fs_list) -> ErrorOr<void> {
+        TRY(m_mounts.with([&](auto& mounts) -> ErrorOr<void> {
+            for (auto& mount : mounts) {
+                if (&mount.guest() != &guest_inode)
+                    continue;
+                auto mountpoint_path = TRY(mount.absolute_path());
+                if (custody_path->view() != mountpoint_path->view())
+                    continue;
+                NonnullRefPtr<FileSystem> fs = mount.guest_fs();
+                TRY(fs->prepare_to_unmount());
+                fs->mounted_count({}).with([&](auto& mounted_count) {
+                    VERIFY(mounted_count > 0);
+                    if (mounted_count == 1) {
+                        dbgln("VirtualFileSystem: Unmounting file system {} for the last time...", fs->fsid());
+                        m_file_systems_list.with([&](auto& list) {
+                            list.remove(*fs);
                         });
+                        if (fs->is_file_backed()) {
+                            dbgln("VirtualFileSystem: Unmounting file backed file system {} for the last time...", fs->fsid());
+                            auto& file_backed_fs = static_cast<FileBackedFileSystem&>(*fs);
+                            file_backed_fs_list.remove(file_backed_fs);
+                        }
+                    } else {
+                        mounted_count--;
                     }
-                } else {
-                    mounted_count--;
-                }
-            });
-            dbgln("VirtualFileSystem: Unmounting file system {}...", fs->fsid());
-            mount.m_vfs_list_node.remove();
-            // Note: This is balanced by a `new` statement that is happening in various places before inserting the Mount object to the list.
-            delete &mount;
-            return {};
-        }
-        dbgln("VirtualFileSystem: Nothing mounted on inode {}", guest_inode.identifier());
-        return ENODEV;
+                });
+                dbgln("VirtualFileSystem: Unmounting file system {}...", fs->fsid());
+                mount.m_vfs_list_node.remove();
+                // Note: This is balanced by a `new` statement that is happening in various places before inserting the Mount object to the list.
+                delete &mount;
```

Comment:
Should that be a FIXME? Seems like we're missing a smart pointer here...

Reply:
We can't use a smart pointer (i.e. `OwnPtr`) here. There was a discussion a few months ago about using Intrusive lists with `OwnPtr` and the conclusion was that a big change to how either the IntrusiveList class works, or introducing a new one (called `IntrusiveListOwnObjects` or something like this) to do that.
Also, this code seems to be new, but it's just that the diff is a bit screwed up and this code was already there.